File: PYTHON/200-299/206_Reverse_Linked_List.py
#Reverse a singly linked list.
#
#click to show more hints.
#
#Hint:
#A linked list can be reversed either iteratively or recursively. Could you implement both?
#
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head == None: return []
        list = []
        node = head
        while node != None:
            list.append(node.val)
            node = node.next
        node = head
        # print list
        for i in list[::-1]:
            node.val = i
            node = node.next
        return head
            

File: Java/200-299/206_ReverseLinkedList.java
//Reverse a singly linked list.

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode p = head;
        if(head == null) return head;
        if(head.next == null)return head;
        ListNode q = head.next;
        head.next = null;
        while(q.next != null){
            ListNode r = q.next;
            q.next = p;
            p = q;
            q = r;
        }
        q.next = p;
        return q;
    }
}
